### PR TITLE
Adding the Add_numbers function

### PR DESCRIPTION
**Pull request by Melat-arch**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by fevziismailsahin** [comment]
You’ve done an excellent job with this implementation! The code is well-organized, the docstrings are clear, and your test coverage is solid.

File: solutions/add_numbers.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    >>> add_numbers(-1, 1)
+    0
+    >>> add_numbers(0.5, 0.5)
+    1.0
+    """
+
+    if not isinstance(a, (int, float)) or not isinstance(b, (int, float)):
+        raise TypeError("Both arguments must be numbers.")
+
+    return a + b
```
Comment (fevziismailsahin) [inline solutions/add_numbers.py:39]: Great job on this I think you’ve done an excellent job. The structure is clear, and the logic is sound. Keep up the great work—this is definitely impressive!”

File: solutions/tests/test_add_numbers.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            add_numbers("a", 1)
+
+    def test_type_error_with_string_second(self):
+        """Test that a TypeError is raised for non-numeric input as second argument."""
+        with self.assertRaises(TypeError):
+            add_numbers(1, "b")
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (fevziismailsahin) [inline solutions/tests/test_add_numbers.py:64]: Great job on writing these unit tests! You’ve covered a good range of test cases, from regular cases to edge and error cases.

File: solutions/tests/test_add_numbers.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+Created on: January 6, 2025
+@author: Melat Assefa
+"""
+
+import unittest
+from solutions.add_numbers import add_numbers
+
+
+class TestAddNumbers(unittest.TestCase):
```
Comment (fevziismailsahin) [inline solutions/tests/test_add_numbers.py:18]: Test for none input. It might be useful to test the function with None as an argument. This can help ensure the function handles unexpected inputs gracefully

**Review by fevziismailsahin — APPROVED** [review]
Great job on the tests. You covered a lot of different cases, and everything looks good. I like how you thought of normal cases and edge cases. A few small ideas for improvement. Try testing with None as input to see if it works. Test with big numbers to make sure there is no mistake with the result. It could be good to check what happens when you add an integer and a float together.
Overall, great work! Just a few small changes to make sure it handles even more cases.
